symfit/core/models.py

```diff
@@ -18,39 +18,49 @@
 else:
     import funcsigs as inspect_sig
 
-def variabletuple(typename, variables, *args, **kwargs):
-    """
-    Create a :func:`~collections.namedtuple` using :class:`~symfit.core.argument.Variable`'s
-    whoses names will be used as `field_names`.
 
-    The main reason for using this object is the `_asdict()` method: whereas a
-    ``namedtuple`` initiates such an :class:`collections.OrderedDict` with the
-    ``field_names`` as keys, this object returns a
-    :class:`collections.OrderedDict` which immediately has the ``Variable``
-    objects as keys.
+class ModelOutput(tuple):
+    def __new__(self, variables, output):
+        """
+        ``variables`` and ``output`` need to be in the same order!
 
-    Example::
+        :param variables: The variables corresponding to ``output``.
+        :param output: The output of a call which should be mapped to
+            ``variables``.
+        """
+        return tuple.__new__(ModelOutput, output)
+
+    def __init__(self, variables, output):
+        """
+        ``variables`` and ``output`` need to be in the same order!
+
+        :param variables: The variables corresponding to ``output``.
+        :param output: The output of a call which should be mapped to
+            ``variables``.
+        """
+        self.variables = variables
+        self.output = output
+        self.output_dict = OrderedDict(zip(variables, output))
+        self.variable_names = {var.name: var for var in variables}
+
+    def __getattr__(self, name):
+        try:
+            var = self.variable_names[name]
+        except KeyError as err:
+            raise AttributeError(err.msg)
+        return self.output_dict[var]
+
+    def __getitem__(self, key):
+        return self.output[key]
+
+    def __str__(self):
+        return self.__class__.__name__ + '(variables={}, output={})'.format(self.variables, self.output)
 
-        >>> x = Variable('x')
-        >>> Result = variabletuple('Result', [x])
-        >>> res = Result(5.0)
-        >>> res._asdict()
-        OrderedDict((x, 5.0))
-
-    :param typename: Name of the `variabletuple`.
-    :param variables: List of :class:`~symfit.core.argument.Variable`, to be used
-        as `field_names`
-    :param args: See :func:`~collections.namedtuple`
-    :param kwargs: See :func:`~collections.namedtuple`
-    :return: Type ``typename``
-    """
     def _asdict(self):
-        return OrderedDict(zip(variables, self))
+        return self.output_dict
```

Comment:
I kept this because this is a method of namedtuple that we use in our codebase. The .copy is a good idea, or we do away with the `_asdict` in our code.